notes-0407-composition-structure.tex

```diff
@@ -0,0 +1,303 @@
+\end{prooftree*}
+
+\begin{prooftree*}
+	\hypo{\oftype{M}{\dfuntype{x}{A}{B}}}
+	\hypo{\oftype{r}{\mathbb{I}}}
+	\infer2{\eqterm{\mathtt{transp}^i\,\parens{\dfuntype{x}{A}{B}}\,r\,M}
+	               {\lam{x}[A]{\mathtt{transp}^i\,B[x \mapsto \text{\texttt{transp-fill}}^{\texttildelow i}\,A\,r\,x]\,r\,\parens{\app{M}{\parens{\text{\texttt{transp-fill}}^1\,A\,r\,x}}}}}
+	               {\dfuntype{x}{A}{B}}}
+\end{prooftree*}
+
+The intuition here essentially follows from function extensionality: we just need the term to have identical input-output behavior. It's therefore perfectly fine to just ``push the operator in,'' instead operating on $B$.
+
+\subsection{Sigma}
+
+\begin{prooftree*}
+	\hypo{\oftype{M}{\dpairtype{x}{A}{B}}}
+	\hypo{\oftype{\varphi,\oftype{i}{\mathbb{I}}}{N}{\dpairtype{x}{A}{B}}}
+	\infer2{\eqterm{\mathtt{hcomp}^i\,\parens{\dpairtype{x}{A}{B}}\,[\varphi\mapsto N]\,M}
+	               {\pair{\mathtt{compFst}}{\mathtt{compSnd}}}
+	               {\dpairtype{x}{A}{B}}}
+\end{prooftree*}
+
+where
+
+\begin{align*}
+	\mathtt{compFst} & \defeq \mathtt{hcomp}^i\,A\,[\varphi\mapsto\pairfst{N}]\,\parens{\pairfst{M}} \\
+	\mathtt{compSnd} & \defeq \mathtt{comp}^i\,B\,[\varphi\mapsto\pairsnd{N}, x\mapsto\mathtt{filler}^i]\,\parens{\pairsnd{M}} \\
+	\mathtt{filler}^i & \defeq \mathtt{hfill}^i\,A\,[\varphi\mapsto\pairfst{N}]\,\parens{\pairfst{M}}
+\end{align*}
+
+\begin{prooftree*}
+	\hypo{\oftype{M}{\dpairtype{x}{A}{B}}}
+	\hypo{\oftype{r}{\mathbb{I}}}
+	\infer2{\eqterm{\mathtt{transp}^i\,\parens{\dpairtype{x}{A}{B}}\,\varphi\,M}
+	               {\pair{\mathtt{transpFst}}{\mathtt{transpSnd}}}
+	               {\dpairtype{x}{A}{B}}}
+\end{prooftree*}
+
+where
+
+\begin{align*}
+	\mathtt{transpFst} & \defeq \mathtt{transp}^i\,A\,\varphi\,\parens{\pairfst{M}} \\
+	\mathtt{transpSnd} & \defeq \mathtt{transp}^i\,\parens{B[x\mapsto\mathtt{filler}^i]}\,\varphi\,\parens{\pairsnd{M}} \\
+	\mathtt{filler}^i & \defeq \text{\texttt{transp-fill}}^i\,A\,\varphi\,\parens{\pairfst{M}}
+\end{align*}
+
+\subsection{Paths}
+
+\begin{prooftree*}
+	\hypo{\oftype{P}{\pathtype{j.A}{M}{N}}}
+	\hypo{\oftype{\varphi,\oftype{i}{\mathbb{I}}}{Q}{\pathtype{j.A}{M}{N}}}
+	\infer2{\eqterm{\mathtt{hcomp}^i\,\pathtype{j.A}{M}{N}\,[\varphi\mapsto Q]\,P}
+	               {\lam{j}{\mathtt{body}}}
+	               {\pathtype{j.A}{M}{N}}}
+\end{prooftree*}
+
+where
+
+\begin{align*}
+	\mathtt{body} & \defeq \mathtt{hcomp}^i\,A\,[\varphi\mapsto Q@j, j=0\mapsto M, j=1\mapsto N]\,\parens{P@j}
+\end{align*}
+
+\begin{prooftree*}
+	\hypo{\oftype{P}{\pathtype{j.A}{M}{N}}}
+	\infer1{\eqterm{\mathtt{transp}^i\,\parens{\pathtype{j.A}{M}{N}}\,\varphi\,P}
+	               {\lam{j}{\mathtt{body}}}
+	               {\pathtype{j.A}{M}{N}}}
+\end{prooftree*}
+
+where
+
+\begin{align*}
+	\mathtt{body} & \defeq \mathtt{comp}^i\,[\varphi\mapsto P@j, j=0\mapsto M, j=1\mapsto N]\,\parens{P@j}
+\end{align*}
+
+\subsection{Positive Types}
+
+Coercion simply ``just works'' for inductive types:
```

Comment:
Technically, this would not work for List(A) or A+B. You need to push the coercion inside the constructors.